### ISSUE TITLE
OneHotEncoder blows up for large values

### ISSUE THREAD

**Issue by MatthiasWinkelmann** [yoshoku/rumale#12]
Here's a quick demonstration, adapted from the spec:

```
  it 'does not murder us' do
    x = Numo::Int32[[0, 0, 5999999], [1, 1, 0], [0, 2, 1], [1, 0, 2]]
    y = Numo::Int32[[0, 1, 1]]
    expect(encoder.fit(x).transform(y)).to eq(Numo::Int32[[1, 0, 0, 1, 0, 0, 1, 0, 0]])
  end
```

This return a vector of length 60_000_000 or so. My system actually shows 1.5 TB(!) of RAM being consumed, although that is luckily almost completely virtual. It is, however, impossible to work with  such values.

I was expecting the transformation to interpret the given values as *catgorical*, as it is mentioned in the docs and is the usual practice as far as I can tell.

**Comment by yoshoku** [comment]
Thank you for your report. I would like to consider modifying this OneHotEncoder's behavior. Until the fix is released, you can obtain the transformation by performing  OrdinalEncoder or LabelEncoder before OneHotEncoder.

```ruby
require 'rumale'

x = Numo::Int32[[0, 0, 5999999], [1, 1, 0], [0, 2, 1], [1, 0, 2]]
y = Numo::Int32[[0, 1, 1]]

ord_enc = Rumale::Preprocessing::OrdinalEncoder.new
x_ord = Numo::Int32.cast(ord_enc.fit_transform(x))
y_ord = Numo::Int32.cast(ord_enc.transform(y))

onehot_enc = Rumale::Preprocessing::OneHotEncoder.new
p onehot_enc.fit(x_ord).transform(y_ord)
# Numo::DFloat#shape=[1,9]
# [[1, 0, 0, 1, 0, 0, 1, 0, 0]]
```

or

```ruby
lbl_enc = Rumale::Preprocessing::LabelEncoder.new
x_lbl = lbl_enc.fit_transform(x.flatten).reshape(*x.shape)
y_lbl = lbl_enc.transform(y.flatten).reshape(*y.shape)

onehot_enc = Rumale::Preprocessing::OneHotEncoder.new
p onehot_enc.fit(x_lbl).transform(y_lbl)
# Numo::DFloat#shape=[1,9]
# [[1, 0, 0, 1, 0, 0, 1, 0, 0]]
```

**Comment by MatthiasWinkelmann** [comment]
Thank you for the quick reply. I had indeed noticed LabelEncoder later last night and wasn't quite sure if the behaviour wasn't intended as-is after all, requiring just a documentation fix and maybe  some protectionary validation.

I checked scikit's implementation and as far as I can tell, *that* changed as well rather recently and deprecated [0,0,...1] style.

**Comment by yoshoku** [comment]
The source code of scikit-learn's OneHotEncoder is changed in version 0.20.0 for encoding string categorical features. It seems that the scikit-learn's OneHotEncoder either encodes using the unique values detected automatically as categories or the given categories with Python list.

```python
>>> import sklearn
>>> print(sklearn.__version__)
0.21.3
>>> import numpy as np
>>> from sklearn.preprocessing import OneHotEncoder
>>> a = np.array([[0, 0, 10], [1, 1, 0], [0, 2, 1], [1, 0, 2]])
>>> b = np.array([[0, 1, 1]])
>>> enc = OneHotEncoder(categories='auto', sparse=False)
>>> enc.fit(a).transform(b)
array([[1., 0., 0., 1., 0., 0., 1., 0., 0.]])
>>> enc.categories_[2]
array([ 0,  1,  2, 10])
```

Currently, since the Rumale's OneHotEncoder does not support string categorical features as input data, it is desirable that the behavior of Rumale's OneHotEncoder be the same as the version 0.19 scikit-learn's OneHotEncoder.

The bug fix will be released within a week.

**Comment by yoshoku** [comment]
I have fixed the bug in OneHotEncoder and released new version 0.13.2. However, in the above example you showed, OneHotEncoder consumes memory in internal process. I recommend encoding with LabelEncoder before OneHotEncoder in the example.
